File: plugins/plugin-discord/rust/src/actions/send_message.rs
use async_trait::async_trait;

use super::{ActionContext, ActionResult, DiscordAction};
use crate::error::{DiscordError, Result};
use crate::types::Snowflake;
use crate::DiscordService;

/// Action that sends a message to a Discord channel.
pub struct SendMessageAction;

#[async_trait]
impl DiscordAction for SendMessageAction {
    fn name(&self) -> &str {
        "SEND_MESSAGE"
    }

    fn description(&self) -> &str {
        "Sends a message to a Discord channel. Use this to respond to users or post content in a channel."
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "SEND_DISCORD_MESSAGE",
            "POST_MESSAGE",
            "REPLY",
            "RESPOND",
            "SAY",
            "CHAT",
        ]
    }

    async fn validate(&self, context: &ActionContext) -> Result<bool> {
        let source = context
            .message
            .get("source")
            .and_then(|v| v.as_str())
            .unwrap_or("");

        if source != "discord" {
            return Ok(false);
        }

        Snowflake::new(context.channel_id.clone())?;

        let has_content = context
            .message
            .get("content")
            .and_then(|c| c.get("text"))
            .map(|t| !t.as_str().unwrap_or("").is_empty())
            .unwrap_or(false);

        Ok(has_content)
    }

    async fn handler(
        &self,
        context: &ActionContext,
        service: &DiscordService,
    ) -> Result<ActionResult> {
        let channel_id = Snowflake::new(context.channel_id.clone())?;

        let content = context
            .message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .ok_or_else(|| DiscordError::InvalidArgument("Missing message content".to_string()))?;

        let message_id = service.send_message(&channel_id, content).await?;

        Ok(ActionResult::success_with_data(
            "Message sent successfully",
            serde_json::json!({
                "message_id": message_id.as_str(),
                "channel_id": channel_id.as_str(),
            }),
        ))
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[tokio::test]
    async fn test_validate_valid() {
        let action = SendMessageAction;
        let context = ActionContext {
            message: json!({
                "source": "discord",
                "content": { "text": "Hello, world!" }
            }),
            channel_id: "123456789012345678".to_string(),
            guild_id: Some("987654321098765432".to_string()),
            user_id: "111222333444555666".to_string(),
            state: json!({}),
        };

        assert!(action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_invalid_source() {
        let action = SendMessageAction;
        let context = ActionContext {
            message: json!({
                "source": "telegram",
                "content": { "text": "Hello" }
            }),
            channel_id: "123456789012345678".to_string(),
            guild_id: None,
            user_id: "111222333444555666".to_string(),
            state: json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }

    #[tokio::test]
    async fn test_validate_missing_content() {
        let action = SendMessageAction;
        let context = ActionContext {
            message: json!({
                "source": "discord",
            }),
            channel_id: "123456789012345678".to_string(),
            guild_id: None,
            user_id: "111222333444555666".to_string(),
            state: json!({}),
        };

        assert!(!action.validate(&context).await.unwrap());
    }
}

File: plugins/plugin-discord/typescript/actions/sendMessage.ts
import {
  type Action,
  type ActionExample,
  type ActionResult,
  type Content,
  composePromptFromState,
  type HandlerCallback,
  type HandlerOptions,
  type IAgentRuntime,
  type Memory,
  ModelType,
  parseJSONObjectFromText,
  type State,
} from "@elizaos/core";
import type { TextChannel } from "discord.js";
import { DISCORD_SERVICE_NAME } from "../constants";
import { requireActionSpec } from "../generated/specs/spec-helpers";
import type { DiscordService } from "../service";

const sendMessageTemplate = `You are helping to extract send message parameters.

The user wants to send a message to a Discord channel.

Recent conversation:
{{recentMessages}}

Extract the following:
1. text: The message text to send
2. channelRef: The channel to send to (default: "current" for the current channel)

Respond with a JSON object like:
{
  "text": "The message to send",
  "channelRef": "current"
}

Only respond with the JSON object, no other text.`;

const spec = requireActionSpec("SEND_MESSAGE");

export const sendMessage: Action = {
  name: spec.name,
  similes: spec.similes ? [...spec.similes] : [],
  description: spec.description,
  validate: async (_runtime: IAgentRuntime, message: Memory, _state?: State): Promise<boolean> => {
    return message.content.source === "discord";
  },
  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    state?: State,
    _options?: HandlerOptions,
    callback?: HandlerCallback
  ): Promise<ActionResult | undefined> => {
    const discordService = runtime.getService(DISCORD_SERVICE_NAME) as DiscordService;

    if (!discordService || !discordService.client) {
      await callback?.({
        text: "Discord service is not available.",
        source: "discord",
      });
      return;
    }

    // Use LLM to extract message parameters
    const prompt = composePromptFromState({
      state,
      template: sendMessageTemplate,
    });

    let messageInfo: { text: string; channelRef?: string } | null = null;

    for (let i = 0; i < 3; i++) {
      const response = await runtime.useModel(ModelType.TEXT_SMALL, {
        prompt,
      });

      const parsedResponse = parseJSONObjectFromText(response);
      if (parsedResponse?.text) {
        messageInfo = {
          text: String(parsedResponse.text),
          channelRef: parsedResponse.channelRef ? String(parsedResponse.channelRef) : "current",
        };
        break;
      }
    }

    if (!messageInfo || !messageInfo.text) {
      runtime.logger.debug(
        { src: "plugin:discord:action:send-message" },
        "[SEND_MESSAGE] Could not extract message info"
      );
      await callback?.({
        text: "I couldn't understand what message you want me to send. Please try again with a clearer request.",
        source: "discord",
      });
      return;
    }

    try {
      const stateData = state?.data;
      const room = stateData?.room || (await runtime.getRoom(message.roomId));

      if (!room || !room.channelId) {
        await callback?.({
          text: "I couldn't determine the current channel.",
          source: "discord",
        });
        return;
      }

      let targetChannelId = room.channelId;

      // If a specific channel was referenced (not "current"), try to find it
      if (messageInfo.channelRef && messageInfo.channelRef !== "current") {
        const guild = discordService.client.guilds.cache.first();
        if (guild) {
          const channels = await guild.channels.fetch();
          const targetChannel = channels.find((ch) => {
            if (!ch || !ch.isTextBased()) return false;
            const channelName = ch.name?.toLowerCase() || "";
            const searchTerm = messageInfo?.channelRef?.toLowerCase() || "";
            return (
              channelName === searchTerm ||
              channelName.includes(searchTerm) ||
              ch.id === messageInfo?.channelRef
            );
          });
          if (targetChannel) {
            targetChannelId = targetChannel.id;
          }
        }
      }

      const channel = await discordService.client.channels.fetch(targetChannelId);
      if (!channel || !channel.isTextBased()) {
        await callback?.({
          text: "I can only send messages to text channels.",
          source: "discord",
        });
        return;
      }

      const textChannel = channel as TextChannel;

      // Send the message
      const sentMessage = await textChannel.send(messageInfo.text);

      const response: Content = {
        text: `Message sent successfully.`,
        source: message.content.source,
      };

      runtime.logger.debug(
        {
          src: "plugin:discord:action:send-message",
          messageId: sentMessage.id,
          channelId: targetChannelId,
        },
        "[SEND_MESSAGE] Message sent successfully"
      );

      await callback?.(response);

      return {
        success: true,
        data: {
          messageId: sentMessage.id,
          channelId: targetChannelId,
        },
      };
    } catch (error) {
      runtime.logger.error(
        {
          src: "plugin:discord:action:send-message",
          agentId: runtime.agentId,
          error: error instanceof Error ? error.message : String(error),
        },
        "Error sending message"
      );
      await callback?.({
        text: "I encountered an error while trying to send the message. Please make sure I have the necessary permissions.",
        source: "discord",
      });
    }
  },
  examples: (spec.examples ?? []) as ActionExample[][],
};

export default sendMessage;
